src/integrators/sdirk_integrators.jl

```diff
@@ -89,165 +98,167 @@ end
+function initialize!(integrator, cache::ImplicitEulerCache)
   integrator.kshortsize = 2
-  @unpack k,fsalfirst = cache
-  integrator.fsalfirst = fsalfirst
-  integrator.fsallast = k
+  integrator.fsalfirst = cache.fsalfirst
+  integrator.fsallast = cache.k
   integrator.k = eltype(integrator.sol.k)(integrator.kshortsize)
   integrator.k[1] = integrator.fsalfirst
   integrator.k[2] = integrator.fsallast
-  integrator.f(integrator.t,integrator.uprev,integrator.fsalfirst) # For the interpolation, needs k at the updated point
+  integrator.f(integrator.t, integrator.uprev, integrator.fsalfirst) # For the interpolation, needs k at the updated point
 end
 
-@muladd function perform_step!(integrator,cache::ImplicitEulerCache,repeat_step=false)
+@muladd function perform_step!(integrator, cache::ImplicitEulerCache, repeat_step=false)
   @unpack t,dt,uprev,u,f = integrator
-  @unpack uf,du1,dz,z,k,J,W,jac_config = cache
+  @unpack uf,du1,dz,z,k,b,J,W,jac_config,tmp,atmp,κ,tol = cache
   mass_matrix = integrator.sol.prob.mass_matrix
 
-  uf.t = t
+  # precalculations
+  κtol = κ*tol
 
+  # calculate W
   if has_invW(f)
-    f(Val{:invW},t,uprev,dt,W) # W == inverse W
+    # skip calculation of inv(W) if step is repeated
+    !repeat_step && f(Val{:invW},t,uprev,dt,W) # W == inverse W
   else
-    if !integrator.last_stepfail && cache.newton_iters == 1 && cache.ηold < integrator.alg.new_jac_conv_bound
+    # skip calculation of J if step is repeated
+    if repeat_step || (!integrator.last_stepfail && cache.newton_iters == 1 && cache.ηold < integrator.alg.new_jac_conv_bound)
       new_jac = false
     else # Compute a new Jacobian
       new_jac = true
       if has_jac(f)
         f(Val{:jac},t,uprev,J)
       else
+        uf.t = t
         if alg_autodiff(integrator.alg)
           ForwardDiff.jacobian!(J,uf,vec(du1),vec(uprev),jac_config)
         else
           Calculus.finite_difference_jacobian!(uf,vec(uprev),vec(du1),J,integrator.alg.diff_type)
         end
       end
     end
-    if integrator.iter < 1 || new_jac || abs(dt - (t-integrator.tprev)) > 100eps()
+    # skip calculation of W if step is repeated
+    if !repeat_step && (integrator.iter < 1 || new_jac || abs(dt - (t-integrator.tprev)) > 100eps())
       new_W = true
       for j in 1:length(u), i in 1:length(u)
-          @inbounds W[i,j] = mass_matrix[i,j]-dt*J[i,j]
+        @inbounds W[i,j] = mass_matrix[i,j]-dt*J[i,j]
       end
     else
       new_W = false
     end
   end
 
+  # initial guess
   if integrator.alg.extrapolant == :linear
-    @. z = integrator.fsalfirst*dt
+    @. z = dt*integrator.fsalfirst
   else # :constant
-    @. z = zero(u)
+    z .= zero(u)
   end
 
-  iter = 0
-  κ = cache.κ
-  tol = cache.tol
-
-  iter += 1
-  f(t+dt,u,k)
-  scale!(k,dt)
+  # initial step of Newton iteration
+  iter = 1
+  tstep = t + dt
+  @. u = uprev + z
+  f(tstep,u,k)
   if mass_matrix == I
-    k .-= z
+    @. b = dt*k - z
```

Comment:
It's necessary if one wants to support units, since `k` is a rate but `dt*k - z` has the same units as `u`.